feat(theme) theme development (#725)
bonitasoft/bonita-doc: md/customize-living-application-theme.md

# Customize the living application theme

::: info
**Note:** For Enterprise, Performance, Efficiency, and Teamwork editions only.
:::

This page explains how to create a theme project in the Studio.

A theme is an archive of files used to define and share CSS definitions and images for an application.
It enables you to specify the same style for all pages and layout of an application.
Each application can have its own theme.

## Prerequisites

> * An Internet connection to fetch maven and npm repositories

Reviewer: Maven

> * Maven basics

Reviewer: ```suggestion
* Maven basic knowledge
```

> * CSS basics

Reviewer: ```suggestion
* CSS basic knowledge
```

> ## Create a Theme project

Reviewer: ```suggestion
## Create a theme project
```

> When you create a new theme (Right click in Project Explorer > New > Theme...), you start with a default project layout based on [Sass](https://sass-lang.com/) preprocessor and SCSS syntax.

Reviewer: ```suggestion
When you create a new theme (right click in Project Explorer > New > Theme...), you start with a default project theme based on [Sass](https://sass-lang.com/) preprocessor and SCSS syntax.
```

::: info

> **Note:** Sass is a css preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic css 3 in a .scss file.

Reviewer: ```suggestion
**Note:** Sass is a CSS preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic CSS 3 in a .scss file.
```

> **Note:** Sass is a CSS preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic CSS 3 in a .scss file.

Reviewer 2: common usage is to write CSS3 without whitespace

:::

> ### Default project layout

Reviewer: ```suggestion
### Default project theme
```

> ### Default project theme

Reviewer 2: I would rather say "Default theme project"

```
myCustomTheme
|__ dist //Contains all resources to package
|	|__ fonts
|	|__ icons
|	|__ images
|	| theme.css //The css file built from scss sources and optimized, do not modify this file directly
|__ node //Node and npm binaries retrieved after first build
|__ node_modules //Npm modules dependencies retrieved after first build
|__ src
|	|__ scss //Contains all your scss source files
|   		| _bonita_pager.scss
|		| _bonita_variables.scss //Contains all the overridden variables for the Bonita theme
|		| main.scss //Main entry point
|__ target //Maven output folder, contains the theme archive after a build
|__ test
|	|__ css
|	|__ js
|	| README.md
|	| index.html //A static test page to preview your theme
| content.xml //Maven assembly descriptor
| package-lock.json //NPM dependencies version lock, ensure a consistent install and compatible dependencies
| package.json //NPM descriptor
| page.properties //Custom page descriptor
| pom.xml //Maven project descriptor
```

> This project is using [Sass](https://sass-lang.com/) to easily create and customize a bootstrap *3.3.7* theme.

Reviewer: ```suggestion
This project is using [Sass](https://sass-lang.com/) to easily create and customize a Bootstrap *3.3.7* theme.
```

> To compile the scss it relies on [Node](https://nodejs.org/en/) and [NPM](https://www.npmjs.com/). See the package.json file below for more detailed.

Reviewer: ```suggestion
To compile the SCSS, it relies on [Node](https://nodejs.org/en/) and [npm](https://www.npmjs.com/). See the package.json file below for more details.
```

> To compile the scss it relies on [Node](https://nodejs.org/en/) and [NPM](https://www.npmjs.com/). See the package.json file below for more detailed.

Reviewer 2: more detail**s**

`package.json`
```json
{
  "name": "myCustomTheme", 
  "version": "0.0.1",
  "description": "My custom theme description",
  "license": "GPL-2.0-or-later",
  "scripts": {
    "build": "node-sass --precision 8 --output-style compressed --omit-source-map-url true --include-path ./node_modules/bootstrap-sass/assets/stylesheets/ src/scss/main.scss target/theme.noprefix.css && postcss target/theme.noprefix.css --no-map --use autoprefixer -b \"last 2 versions\" -o dist/theme.css"
  },
  "devDependencies": {
    "node-sass": "4.11.0",
    "postcss-cli": "6.1.2",
    "autoprefixer": "9.5.0",

>     "bootstrap-sass": "3.3.7" //Supported version of bootstrap in Bonita

Reviewer: ```suggestion
    "bootstrap-sass": "3.3.7" //Supported version of Bootstrap in Bonita
```

  }
}
```

By default, a _build_ npm script is defined. It runs `node-sass` to compile the `src/scss/main.scss` file. The build command includes the bootstrap-sass stylesheets in order to have clean `@import` statements in the _*.scss_ files.
In addition, *postcss-cli* and *autoprefixer* are used to add vendor prefixes for a better browser compatibility.

> The maven descriptor is responsible to run the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more detailed.  

Reviewer: ```suggestion
The Maven descriptor is responsible for running the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more details.  
```

> The maven descriptor is responsible to run the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more detailed.  

Reviewer 2: responsible for running the npm
more details

		<plugins>
			<plugin>
				<groupId>com.github.eirslett</groupId>
				<artifactId>frontend-maven-plugin</artifactId>
				<executions>
					<execution>
						<id>install node and npm</id>
						<goals>
							<goal>install-node-and-npm</goal>
							<goal>npm</goal>
						</goals>
					</execution>
					<execution>
						<id>npm build</id>
						<goals>
							<goal>npm</goal>
						</goals>
						<phase>prepare-package</phase>
						<configuration>
							<arguments>run build</arguments>
						</configuration>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-assembly-plugin</artifactId>
				<executions>
					<execution>
						<id>page-content</id>
						<phase>package</phase>
						<goals>
							<goal>single</goal>
						</goals>
						<inherited>false</inherited>
						<configuration>
							<ignoreDirFormatExtensions>true</ignoreDirFormatExtensions>
							<appendAssemblyId>false</appendAssemblyId>
							<descriptors>
								<descriptor>content.xml</descriptor>
							</descriptors>
						</configuration>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
</project>
```

> The `artifactId`, `name` and `description` are used to defined the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and NPM version are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have  NodeJS and NPM already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more informations.

Reviewer: ```suggestion
The `artifactId`, `name` and `description` are used to define the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and npm versions are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have NodeJS and npm already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more information.
```

> The `artifactId`, `name` and `description` are used to defined the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and NPM version are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have  NodeJS and NPM already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more informations.

Reviewer 2: used to defin**e**

### SCSS source files

`src/scss/main.scss`
```css
//Bonita variables
@import "bonita_variables";

// Bootstrap
@import "bootstrap";

@import "bonita_pager";
```
The main.scss is the aggregation of 3 imports:

> * `@import "bonita_variables";` import the content of `src/scss/_bonita_variables.scss` file.

Reviewer: ```suggestion
* `@import "bonita_variables";` imports the content of `src/scss/_bonita_variables.scss` file.
```

> * `@import "bootstrap";` import the bootstrap-sass stylesheet. You may look its content in `node_modules/bootstratp-sass/assets/stylesheets/_bootstrap.scss`.

Reviewer: ```suggestion
* `@import "bootstrap";` imports the bootstrap-sass stylesheet. You may look its content in `node_modules/bootstratp-sass/assets/stylesheets/_bootstrap.scss`.
```

> * `@import "bonita_pager";` import the content of `src/scss/_bonita_pager.scss` file, a custom style for bootstrap pager used by the Bonita theme. 

Reviewer: ```suggestion
* `@import "bonita_pager";` imports the content of `src/scss/_bonita_pager.scss` file, a custom style for Bootstrap pager used by the Bonita theme. 
```

> When using Sass you can split your stylesheets into _partials_. This is a great way to modularize your CSS and help keep things easier to maintain. A partial is simply a Sass file named with a leading underscore. You might name it something like `_partial.scss`. The underscore lets Sass know that the file is only a partial file and that it should not be generated into a CSS file. Sass partials are used with the `@import` directive like in our `src/scss/main.scss`. 

Reviewer: ```suggestion
When using Sass, you can split your stylesheets into _partials_. This is a great way to modularize your CSS and help keep things easier to maintain. A partial is simply a Sass file named with a leading underscore. You might name it something like `_partial.scss`. The underscore lets Sass know that the file is only a partial file and that it should not be generated into a CSS file. Sass partials are used with the `@import` directive like in our `src/scss/main.scss`. 
```

$dropdown-link-hover-color: $white;
$dropdown-link-hover-bg: #2c3e50;

//Pagination
$pagination-color: $white;
$pagination-bg: $brand-primary;
$pagination-border:  transparent;

$pagination-hover-color: $pagination-color;
$pagination-hover-bg: darken($brand-primary, 15%);
$pagination-hover-border: transparent;

$pagination-active-bg: darken($brand-primary, 15%);
$pagination-active-border: transparent;

$pagination-disabled-color: #ecf0f1;
$pagination-disabled-bg: #476481;
$pagination-disabled-border: transparent;

//Pager
$pager-color: $pagination-color;
$pager-bg: $brand-primary;
$pager-hover-color: $pagination-hover-color;

//Badge
$badge-bg: $brand-primary;

//Panel
$panel-border-color: #dddddd;
$panel-primary-border: $panel-border-color;
$panel-success-border: $panel-border-color;
$panel-info-border: $panel-border-color;
$panel-warning-border: $panel-border-color;
$panel-danger-border: $panel-border-color;

$panel-success-heading-bg: $brand-success;
$panel-info-heading-bg: $brand-info;
$panel-warning-heading-bg: $brand-warning;
$panel-danger-heading-bg: $brand-danger;

//Glyphicons fonts
$icon-font-path: "./fonts/"; // path relative to the theme.css file in the dist folder
```

> All the variables defined in this file are used by *bootstrap-sass*. You can look the following scss file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.

Reviewer: ```suggestion
All the variables defined in this file are used by *bootstrap-sass*. You can look at the following SCSS file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.
```

> All the variables defined in this file are used by *bootstrap-sass*. You can look the following scss file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.

Reviewer 2: you can look **at** the following

::: info
**Note:** Only variables declared with the `!default` flag can be overridden.
:::

> SASS and SCSS have a lots of other interesting features that you should use. Check [SASS documentation](https://sass-lang.com/documentation) to known more.

Reviewer: ```suggestion
Sass and SCSS have lots of other interesting features that you could use. Check [Sass documentation](https://sass-lang.com/documentation) to known more.
```

> ## Building, Deploying, Previewing a Theme

Reviewer: ```suggestion
## Building, Deploying, and Previewing a theme
```

> During the development phase you can have preview your theme using the provided test page in `test/index.html`.

Reviewer: ```suggestion
During the development phase, you can preview your theme using the provided test page in `test/index.html`.
```

> First you will have to `build` your theme: Right click on your theme project > Build or press `Ctrl+Shift+B` shortcut.

Reviewer: ```suggestion
First you will have to `build` your theme: right click on your theme project > Build or press `Ctrl+Shift+B` for a shortcut.
```

> If your theme is already associated to a living application you can just `deploy` your theme:  Right click on your theme project > Deploy or press `Ctrl+Shift+D` shortcut.

Reviewer: ```suggestion
If your theme is already associated to a living application, you can just `deploy` your theme:  right click on your theme project > Deploy or press `Ctrl+Shift+D` for a shortcut.
```

> ## Create a Theme project from an existing theme in production

Reviewer: ```suggestion
## Create a theme project from an existing theme in production
```

If you are upgrading from a previous Bonita version you may already have a theme.css file packaged in a custom page archive.  

> The easiest way of integrating your theme as a new Theme project is:

Reviewer: ```suggestion
The easiest way of integrating your theme as a new theme project is:
```

> * Create a New Theme

Reviewer: ```suggestion
* 1. Create a new theme
```

> * Extract the theme.css file from your Theme custom page zip

Reviewer: ```suggestion
* 2. Extract the theme.css file from your theme custom page .zip archive
```

> * Replace the content of the main.scss file with the content of the extracted theme.css file

Reviewer: ```suggestion
* 3. Replace the content of the main.scss file with the content of the extracted theme.css file
```

> * Retrieve all the related assets if any and copy them in the `dist` folder accordingly

Reviewer: ```suggestion
* 4. Retrieve all the related assets if any and copy them in the `dist` folder accordingly
```
